Fixing LU pivot bugs

AtheMathmo opened the pull request:
This resolves issue #74 (pending approval from @bright-star ).

There was an issue with the previous pivoting - often times the pivot matrix would not actually pivot correctly (while still producing a valid orthogonal matrix and so not failing in every case). The new pivot technique works directly on the columns which may be slightly less efficient (we also remove a transpose) - but is _more_ correct.

Andlon commented:
Could you perhaps also include a test of an example that failed with the old implementation, so that it we can prevent future regressions?

AtheMathmo commented:
Thanks for reminding me!

I added a regression test for `solve` - which is what reported the error in #74. I also added a sanity test for the LUP decomp of the matrix in question, just to check it doesn't error (as it did before).

bright-star commented:
I'll try this ASAP, thanks!

bright-star commented:
No more errors for me! I haven't numerically verified it yet, though.

AtheMathmo commented:
Great! I numerically verified the example you gave in the issue. Let me know if you want to additional checks - if so I'll merge afterwards.
